added Units for AtomsBase interface + a Base.show routine

JanHab opened the pull request:
Hello - I added Units for the box and positions.

In AtomsBase the user can use any length unit and is not forced to use Ängstrom. Therefore I added a _write_convert routine specified for Length units which converts the units in Ängstrom and then uses ustrip. to cut the unit.
I would like to add this for units like mass and velocity, if this is wished by you, but could not find the units ExtXYZ uses in this cases. At the moment these units are just cutted of.

Additionally I added that the atomic_masses are looked up in the Atoms function and are deleted from atom_data in the write_dict function so that they are not written in the final file.
Is this wished or should they be written down in the file as well?

jameskermode commented:
Thanks, this is nice. I agree it would be good to add unit conversion for velocities and masses. ExtXYZ spec doesn’t impose a choice of time units, but let’s follow the ASE units system:

https://wiki.fysik.dtu.dk/ase/ase/units.html

which means lengths are in A, energies in eV and masses are in amu and time thus has somewhat strange units.

jameskermode commented:
Regarding the atomic_masses property, I agree with removing them from the dict to skip writing but only if they were automatically created from chemical species. If the input file had a custom mass property this should be persevered in a read / write / read round trip. Not sure how best to implement that.

jameskermode commented:
Please could you also add a test of the unit conversion functionally?

JanHab commented:
atomic_masses: 
What do you think about getting the atomic_mass of the atom type with PeriodicTable and comparing this to the atomic_mass of the atom. If both are not the same the atomic_masses are kept and written in the file.
If the input file has atomic_masses aren't they already read and saved as atomic_masses in atom_data?

unit test:
Isn't his already tested in the AtomsBase testset? There is read the infile (seq1), saved in a new file and read again (seq2). Both are tested if they are the same. If the unit conversion would fail, wouldn't we get a failed test there?
I can add an extra test were it is tested if the unit is Ängstrom e.g. if you would like to have this?

jameskermode commented:
Both solutions sound good including the extra test of unit conversion, please go ahead.

JanHab commented:
atomic_masses:
Now the atomic_mass is kept if this differs to the atomic_mass got by PeriodicTable.
If the input file got a custom mass property this atomic_masses are read with the unit "u".

unit test:
I added a unit test for position, bounding_box and atomic_mass. 
Just the position, bounding_box and atomic_mass are read with units yet. However, energies and other properties are read without any unit.
I am not sure if this should be changed.
I could add units for energies like dft_energy, but i am not sure if there usually occur a lot of different energies in .extxyz files?
This would lead to a lot of extra cases to add units to all of these energy types.
They still would be converted in "eV" when a system is saved to a file.

What are your thoughts on this?

jameskermode commented:
I think also converting units for the ASE default property names of `momenta` (from `u*A*(A*sqrt(u/eV)`), `energy` and `energies` (from eV), `forces` (from eV/A), `stress` and `stresses` (from eV/A^3) and `virial` (eV) would also be nice, but as you point out there could often by multiple energies, forces etc contained within an extxyz file, and I don't think it's sensible to try to infer the users' intentions from their choice of name.

JanHab commented:
Okay, now i have added those units, hopefully i don't miss any unit?
I added test routines for all of those units where i create a system, save this, load it again and check that the unit conversion was right. This seems to work.

For velocities I have added that they also get a unit by reading, but are not set on zero by default, as you wished.

I have added the Base.show routine with the name "Atoms" for the system. There i am not sure how to create a test for this, any ideas?

The data routine I have skipped for the moment, I think i understood something wrong about the data field in AtomsBase. I will think about how to do this the best. In AtomsBase there is no way to get all this information, saved in the data block of FlexibleSystem or Atom, yet, but i think it would be nice to add this in both, AtomsBase and here. 

What are your thoughts on this? Happy about any feedback or further ideas.

JanHab commented:
Okay I have added a test for the show representation.
Not sure if this is exactly what you meant, but it tests if the representation of seq1 is equal to the representation of seq2.
It works with a String comparison.

jameskermode reviewed (commented):
Looking good generally, just a couple of minor comments

jameskermode commented on test/runtests.jl:
```diff
@@ -121,6 +121,7 @@ Si        13.00000000      14.00000000      $(frame+1).00000000          0
             ExtXYZ.save(outfile, seq1)
             seq2 = ExtXYZ.load(outfile)
             @test all(seq1 .≈ seq2)
+            @test all(repr(seq1) == repr(seq2))
```
I had more in mind to do something like `@test eval(Meta.parse(repo(seq1))) ≈ seq1`, but this may not work with the output formats used in AtomsBase

JanHab replied:
I just tried it and it seems not to work, in the case for your testfile it says 

> Si not defined

I think the atomic_symbols are a problem here.

jameskermode replied:
OK, that's fine then.

JanHab commented on src/atoms.jl:
```diff
@@ -160,3 +201,4 @@ end
 save(file::Union{String,IOStream}, system::Atoms; kwargs...) = write_frame(file, write_dict(system); kwargs...)
 
 save(file::Union{String,IOStream}, systems::Vector{Atoms{NT1,NT2}}; kwargs...) where {NT1,NT2} = write_frames(file, write_dict.(systems); kwargs...)
+save(file::Union{String,IOStream}, systems::Vector{Atoms}; kwargs...) = write_frames(file, write_dict.(systems); kwargs...)
```
I ran into a problem when i loaded the test.xyz file and tried to safe this.
Then this error occured:
> ERROR: MethodError: no method matching save(::String, ::Vector{Atoms})
Closest candidates are:
  save(::Union{IOStream, String}, ::Atoms; kwargs...) at C:\Users\janha\Documents\Programming\git\SEP\ExtXYZ.jl\src\atoms.jl:201
  save(::Union{IOStream, String}, ::Array{Atoms{NT1, NT2}, 1}; kwargs...) where {NT1, NT2} at C:\Users\janha\Documents\Programming\git\SEP\ExtXYZ.jl\src\atoms.jl:203
Stacktrace:
 [1] top-level scope
   @ REPL[5]:1

I do not fully understand the reason of this error, because in the tests everything seems to work, however with this extra method the error didn't occur.
Complete forgot that i changed this, sorry for no comment on this.

jameskermode replied:
OK, let's go with this.

JanHab commented on src/atoms.jl:
```diff
@@ -51,6 +51,27 @@ _read_convert(value::Int32) = Int(value)
 
 _write_convert(value) = value
+_write_convert(value::T) where {T<:Unitful.Length} = ustrip(uconvert(u"Å", value))
+_write_convert(value::Vector{T}) where {T<:Unitful.Length} = ustrip.(uconvert.(u"Å", value)) 
```
I agree that all of these routines wouldn't be needed in the case by calling it `_write_convert.(value)`.
This should work, but i don't know how to call it in (l. 83):
`_dict_remap_rev(d) = _dict_remap_names(d, REV_NAME_MAP, string, String, _write_convert)`

I even thought about adding a routine for a Matrix, because yet there is no possibility to use units with a Matrix (just with Vector{Vector{T}}, but i agree, if one calls it with `_write_convert.(value)` there should be no need for all of these routines.

JanHab replied:
I think those two routines are needed, because otherwise the hcat((value)...) part would be lost in the case it is needed.

jameskermode replied:
OK, good solution

jameskermode commented:
Thanks for all the hard work. I'll go ahead and merge and then tag a new release. We can pick up any further things in new PR(s).

JanHab commented:
Cool, thanks a lot.

jameskermode commented:
@JanHab v0.1.7 has been released via General registry
